Added more code style rules

penev92 opened the pull request:
This is a continuation of my recent efforts to finally have our toolchain enforce our long-standing code style rules.
Follow-up of  #19747 and #19801, this adds ~~two~~ three more rules to `.editorconfig` to treat ~~two~~ three more suggestions as warnings, which the CI then treats as errors.
Calling @RoosterDragon.

penev92 committed:
Fixed unnecessary zero-length array allocations

Changed all currently present zero-length array allocations in the codebase to use `Array.Empty` instead.

penev92 committed:
Added a rule about zero-length array allocations

penev92 committed:
Fixed inconsistent declaration modifier order

penev92 committed:
Added a rule about unordered modifiers

pchote commented on .editorconfig:
```diff
@@ -111,6 +111,9 @@ dotnet_style_require_accessibility_modifiers = omit_if_default:warning
 #raise a warning on build when default access modifiers are explicitly specified
 dotnet_diagnostic.IDE0040.severity = warning
 
+dotnet_diagnostic.CA1825.severity = warning    # Avoid unnecessary zero-length array allocations.
```
The rest of the file places comments on the line before the rule. Can we please keep a consistent style?

penev92 replied:
To be honest I would really like to move away from the current style in this file and towards something more like the style in `OpenRA.ruleset`, where everything is neat and tidy, while here we have several comment rule violations - no empty line above, no leading space and no leading capital letter 😐 
But we should probably discuss the still-existing separation between style rules in `.editorconfig` and in `OpenRA.ruleset`.

pchote replied:
Sure, I have no problem with changing all the other comments to match this - but lets keep it consistent.

penev92 commented:
To resolve the issue of the style inside `.editorconfig` I'm making this depends on #19873.

RoosterDragon commented on .editorconfig:
```diff
@@ -84,35 +82,27 @@ dotnet_naming_rule.parameters.severity = warning
-dotnet_style_require_accessibility_modifiers = omit_if_default:warning
-#raise a warning on build when default access modifiers are explicitly specified
-dotnet_diagnostic.IDE0040.severity = warning
+dotnet_diagnostic.IDE0011.severity = none                                # Don't prefer braces (for one liners).
```
Would prefer `silent` rather than `none`. This allows the IDE to offer the refactoring if you specifically invoke for possible fixes on that line, and it still won't show up in the issues list. `none` turns off the ability to see that option at all.

I tend to use `none` for cases such as outright broken rules or rules that are sufficiently likely to be dangerous for somebody unfamiliar with them you don't want them appearing at all. For things that violate style but are reasonable I'm-working-on-this refactorings (e.g. here if you wanted to add more statements to a one-line branch) then `silent` is my go to level.

penev92 replied:
We specifically want to have no braces in these cases, but alright, as long as the IDE doesn't *suggest* adding them then I'm fine with it.

RoosterDragon reviewed (dismissed):
LGTM. One fixup request.

penev92 committed:
Changed one-liner braces rule to silent

Having this set to "none" disabled the IDE's option to add braces, whereas "silent" lets it do it on the user's request while still not suggesting it on its own.

penev92 committed:
Added more explanations to .editorconfig

penev92 commented:
Added two more commits - one to change `dotnet_diagnostic.IDE0011.severity` from `none` to `silent` as requested and one to add more comments to `.editorconfig` to give context for some of the rules there.

penev92 commented:
Rebased. Only changes are to the comments in `.editorconfig`.

penev92 committed:
Reordered code style rules

To order them by their number.

abcdefg30 reviewed (approved):
Ok.
